core/src/main/scalajvm/sttp/client4/internal/httpclient/MultipartBodyBuilder.scala

```diff
@@ -0,0 +1,152 @@
+import sttp.client4.internal.Utf8
+import sttp.monad.MonadError
+import sttp.monad.syntax._
+
+import java.net.http.HttpRequest
+import java.util.function.Supplier
+import java.io.File
+import java.io.InputStream
+import java.nio.Buffer
+import java.nio.ByteBuffer
+import java.io.ByteArrayInputStream
+import java.util.UUID
+import java.nio.charset.StandardCharsets
+import scala.collection.JavaConverters._
+
+trait MultipartBodyBuilder[BinaryStream, F[_]] {
+  def multipartBodyPublisher(
+      parts: Seq[Part[GenericRequestBody[_]]]
+  )(implicit m: MonadError[F]): (F[HttpRequest.BodyPublisher], String)
+}
+
+class NonStreamMultipartBodyBuilder[BinaryStream, F[_]] extends MultipartBodyBuilder[BinaryStream, F] {
+  override def multipartBodyPublisher(
+      parts: Seq[Part[GenericRequestBody[_]]]
+  )(implicit m: MonadError[F]): (F[HttpRequest.BodyPublisher], String) = {
+    val multipartBuilder = new MultiPartBodyPublisher()
+    parts.foreach { p =>
+      val allHeaders = Header(HeaderNames.ContentDisposition, p.contentDispositionHeaderValue) +: p.headers
+      val partHeaders = allHeaders.map(h => h.name -> h.value).toMap.asJava
+      p.body match {
+        case NoBody         => // ignore
+        case FileBody(f, _) => multipartBuilder.addPart(p.name, f.toFile.toPath, partHeaders)
+        case StringBody(b, e, _) if e.equalsIgnoreCase(Utf8) => multipartBuilder.addPart(p.name, b, partHeaders)
+        case StringBody(b, e, _) =>
+          multipartBuilder.addPart(p.name, supplier(new ByteArrayInputStream(b.getBytes(e))), partHeaders)
+        case ByteArrayBody(b, _) =>
+          multipartBuilder.addPart(p.name, supplier(new ByteArrayInputStream(b)), partHeaders)
+        case ByteBufferBody(b, _) =>
+          if ((b: Buffer).isReadOnly)
+            multipartBuilder.addPart(p.name, supplier(new ByteBufferBackedInputStream(b)), partHeaders)
+          else
+            multipartBuilder.addPart(p.name, supplier(new ByteArrayInputStream(b.array())), partHeaders)
+        case InputStreamBody(b, _) => multipartBuilder.addPart(p.name, supplier(b), partHeaders)
+        case StreamBody(_) =>
+          throw new IllegalArgumentException("Multipart streaming bodies are not supported with this backend")
+        case m: MultipartBody[_] => throwNestedMultipartNotAllowed
+      }
+    }
+    (multipartBuilder.build().unit, multipartBuilder.getBoundary)
+  }
+
+  private def supplier(t: => InputStream) =
+    new Supplier[InputStream] {
+      override def get(): InputStream = t
+    }
+}
+
+trait StreamMultipartBodyBuilder[BinaryStream, F[_]] extends MultipartBodyBuilder[BinaryStream, F] {
+  def fileToStream(file: File): BinaryStream
+
+  def byteArrayToStream(array: Array[Byte]): BinaryStream
+
+  def concatStreams(stream1: BinaryStream, stream2: BinaryStream): BinaryStream
+
+  def toPublisher(stream: BinaryStream): F[HttpRequest.BodyPublisher]
+
+  override def multipartBodyPublisher(
```

Comment:
maybe let's just call the method `apply`? the repetition doesn't give much more additional info

Reply:
Done